sdks/python/apache_beam/runners/interactive/caching/filebasedcache.py

```diff
@@ -0,0 +1,276 @@
+    if not self._coder_was_provided and self.coder is not None:
+      self.coder = None
+
+  def remove(self):
+    self._finalizer()
+
+  @property
+  def removed(self):
+    return not self._finalizer.alive
+
+  def __del__(self):
+    self.remove()
+
+  @property
+  def file_pattern(self):
+    return self.location + '*'
+
+  @property
+  def _reader_kwargs(self):
+    reader_kwargs = {
+        k: v
+        for k, v in self._writer_kwargs.items()
+        if k in self._reader_passthrough_arguments
+    }
+    return reader_kwargs
+
+  @property
+  def _file_path_prefix(self):
+    return self.location + "-{:03d}".format(self._num_writes)
+
+
+class TextBasedCache(FileBasedCache):
+
+  _reader_class = textio.ReadFromText
+  _writer_class = textio.WriteToText
+  _reader_passthrough_arguments = {"coder", "compression_type"}
+  requires_coder = True
+
+  def __init__(self, location, **writer_kwargs):
+    warnings.warn("TextBasedCache is not reliable and should not be used.")
+    super(TextBasedCache, self).__init__(location, **writer_kwargs)
+
+
+class SafeTextBasedCache(FileBasedCache):
+
+  _reader_class = textio.ReadFromText
+  _writer_class = textio.WriteToText
+  _reader_passthrough_arguments = {"coder", "compression_type"}
+  requires_coder = True
+
+  def __init__(self, location, **writer_kwargs):
+    writer_kwargs["coder"] = SafeFastPrimitivesCoder()
+    super(SafeTextBasedCache, self).__init__(location, **writer_kwargs)
+
+
+class TFRecordBasedCache(FileBasedCache):
+
+  _reader_class = tfrecordio.ReadFromTFRecord
+  _writer_class = tfrecordio.WriteToTFRecord
+  _reader_passthrough_arguments = {"coder", "compression_type"}
+  requires_coder = True
+
+
+class ParquetBasedCache(FileBasedCache):
+
+  _reader_class = parquetio.ReadFromParquet
+  _writer_class = parquetio.WriteToParquet
+  _reader_passthrough_arguments = {}
+  requires_coder = False
+
+
+class AvroBasedCache(FileBasedCache):
+
+  _reader_class = avroio.ReadFromAvro
+  _writer_class = avroio.WriteToAvro
+  _reader_passthrough_arguments = {"use_fastavro"}
+  requires_coder = False
+
+
+class PatchedWriter(PTransform):
+
+  def __init__(self, cache, writer_class, writer_args, writer_kwargs):
+    self._cache = cache
+    self._writer_class = writer_class
+    self._writer_args = writer_args
+    self._writer_kwargs = writer_kwargs
+
+  def expand(self, pcoll):
+    if self._cache.requires_coder and self._cache.coder is None:
+      self._cache.coder = coders.registry.get_coder(pcoll.element_type)
+    writer_kwargs = self._writer_kwargs.copy()
+    if self._cache.requires_coder:
+      writer_kwargs["coder"] = self._cache.coder
+    writer = self._writer_class(*self._writer_args, **writer_kwargs)
+    return pcoll | writer
+
+
+class SafeFastPrimitivesCoder(coders.Coder):
+  """This class add an quote/unquote step to escape special characters."""
+
+  # pylint: disable=deprecated-urllib-function
+
+  def encode(self, value):
+    return quote(
```

Comment:
@qinyeli may be able to provide more feedback, since she is the one who first addressed the issue. 

From my understanding, the `apache_bem.coders.StrUtf8Coder` coder supports writing string data only, while we want to be able to serialize any data format. There is `apache_bem.coders.FastPrimitivesCoder`, but if we write the output of `FastPrimitivesCoder.encode(element)` directly to a file, we lose the information about when a given element stops and a new element starts. 

For example:

```python
temp = io.BytesIO()
coder = beam.coders.FastPrimitivesCoder()

data = [{"a": 11}]

for element in data:
    element_encoded = coder.encode(element)
    print("in", element, element_encoded)
    temp.write(element_encoded)
    
temp.seek(0)
for line in temp.read().split(b"\n"):
    print("out", coder.decode(line), line)
```

prints:

```raw
in {'a': 11} b'\x07\x01\x04\x01a\x01\x0b'
out {'a': 11} b'\x07\x01\x04\x01a\x01\x0b'
```

but the following code:

```python
temp = io.BytesIO()
coder = beam.coders.FastPrimitivesCoder()

data = [{"a": 10}]

for element in data:
    element_encoded = coder.encode(element)
    print("in", element, element_encoded)
    temp.write(element_encoded)
    
temp.seek(0)
for line in temp.read().split(b"\n"):
    print("out", coder.decode(line), line)
```

prints:

```raw
in {'a': 10} b'\x07\x01\x04\x01a\x01\n'
out {'a': 0} b'\x07\x01\x04\x01a\x01'
out None b''
```.

`apache_bem.coders.FastPrimitivesCoder` also fails for numpy arrays.